sklearn/linear_model/_sag_fast.pyx.tp

```diff
@@ -534,6 +536,10 @@ def sag{{name}}(SequentialDataset{{name}} dataset,
             elif verbose:
                 printf('Epoch %d, change: %.8f\n', n_iter + 1,
                                                   max_change / max_weight)
+            if callbacks is not None:
+                with gil:
+                    _eval_callbacks(callbacks, n_iter=n_iter, coef=weights_array,
+                                    intercept=intercept_array)
```

Comment:
I remember a strange behavior during the paris sprint last year (I think it was with @pierreglaser and @tomMoral ?) where the GIL was acquired in a condition like this, and even when the condition was always False the code was significantly slower.

Might be something to keep in mind

Reply:
It's clearly an issues with parallel code https://github.com/cython/cython/issues/3554 (and I'm not sure how to handle parallel code with callbacks so far).

However acquiring GIL in long running loops occasionally is beneficial as users can't interrupt calculation with Ctrl+C otherwise. So acquiring GIL at the end of each epoch would actually solve a bug here https://github.com/scikit-learn/scikit-learn/issues/9136#issuecomment-413140049

Will switch to acquire the GIL at the end of each epoch even if callback is None.